### ISSUE TITLE
CSP improvements

### ISSUE THREAD

**Issue by nijel** [phpmyadmin/phpmyadmin#12437]
Usage of not needed CSP values (result in weaker CSP policy)
 
from ReCaptcha Documentation: https://developers.google.com/recaptcha/docs/faq#im-using-content-security-policy-csp-on-my-website-how-can-i-configure-it-to-work-with-recaptcha
The needed directives are:
        
        script-src https://www.google.com/recaptcha/, https://www.gstatic.com/recaptcha/
        frame-src https://www.google.com/recaptcha/
        style-src 'unsafe-inline'
    
so the below are probably not needed ($captcha_url in style-src & img-src)
    
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L553
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L560
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L572
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L585
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L590
    
and it's better to use frame-src (deprecated) or child-src

*        https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives#child-src
*        https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives#frame-src
        
instead of default-src for $captcha_url in:
    
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L565
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L577
*    https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L590
 
12) CSP Improvements
 
12.1)
add reflected-xss CSP directive to Content-Security-Policy header only (as it's available only from CSP Level 2)
    reflected-xss block
    
12.2)
add object-src CSP directive such as:
    object-src 'none'
in order to disallow plugin usage (embed/object/applet) tags which isn't used by PMA.
 
more information:

*    https://groups.google.com/a/chromium.org/forum/#!msg/security-dev/UqCSmNUHhNg/XBlvV_E5eowJ
    
12.3)
 
add form-action CSP directive such as:
    form-action 'self'
* https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives#form-action
This directive can be used to mitigate:
    XSS by URI Injection [ javascript:alert(1) ] in action attribute
    Leak of data to attacker domain via form submission.

**Comment by nijel** [comment]
It's one of the security improvements suggested by @emanuelb. In short CSP should be as narrow as possible and there are some not necessary directives included.

**Comment by nijel** [comment]
Currently the related code is here:

https://github.com/phpmyadmin/phpmyadmin/blob/78638600dbad45294b87b91d362f148675324598/libraries/classes/Header.php#L521-L606

**Comment by pirland** [comment]
Hello, is this still an active issue that needs work?

**Comment by williamdes** [comment]
@pirland This issue still needs work, you can work on it :)

**Comment by ksimenic** [comment]
Based on the [updated FAQ](https://developers.google.com/recaptcha/docs/faq#im-using-content-security-policy-csp-on-my-website.-how-can-i-configure-it-to-work-with-recaptcha) new recommendation is as follows.
> We recommend using the nonce-based approach documented with CSP3. Make sure to include your nonce in the reCAPTCHA api.js script tag, and we'll handle the rest.

**Comment by pirland** [comment]
Most of these are addressed by my recent pull request:  https://github.com/phpmyadmin/phpmyadmin/pull/15539

12.1 reflected-xss seems to be depreciated.  I don't think it is required any more, but please correct me if I'm wrong.

I changed style-src to use a nonce, but doing so automatically disables unsafe-inline .  This interfered with other inline styles being used, so I disabled it for the time being.

Removing inline styles project wide, and re-enabling the nonce is something I'm willing to tackle, but it seemed like it might best be started as a new issue.

**Comment by pirland** [comment]
Hey, I was hoping someone could take a look at my pull request: https://github.com/phpmyadmin/phpmyadmin/pull/15539
I am pretty new to git, and I seem to have picked up someone else's changes and added them into my fork.  When I made my pull request I was prompted to rebase, in order to sign off on all my commits, which I thought I had already done.  Previous to that I thought I had only  pulled changes from the master branch of the main repository, so I am not sure what specifically I did wrong or how best to undo it.

**Comment by pirland** [comment]
I'm checking back in on this.  Would it perhaps be easier if I created a fresh branch, and integrated my changes into that?  I believe they were confined to one file.

**Comment by williamdes** [comment]
@pirland I you can split your work for CSP in one PR and another for the extra fixes and improvements It would help reviewing and merging ?
thank you for the contribution you made in #15539, I am sure It wan can be simplified to only CSP changes :)

**Comment by rabeehrz** [comment]
@williamdes Can I work on this?

**Comment by williamdes** [comment]
Yes @rabeehrz

**Comment by pirland** [comment]
I know I've been MIA this,  but if it's not too late,  I'd still like to
upload my changes.

Paul

**Comment by williamdes** [comment]
> I know I've been MIA this, but if it's not too late, I'd still like to upload my changes. Paul

Yes, you can upload them but maybe you can merge your work with @rabeehrz
